src/devices/machine/cs8900a.cpp

```diff
@@ -0,0 +1,1522 @@
+  else if(ppaddress<0x140) {
+#ifdef CS8900_DEBUG_WARN_REG
+    log_message(cs8900_log,
+                "WARNING! Ignoring write to read-only Status Register %04x",
+                ppaddress);
+#endif
+    return;
+  }
+
+  /* --- write to initiate transmit register range --- */
+  else if(ppaddress<0x150) {
+    /* check tx_cmd register */
+    if(ppaddress==0x144) {
+      /* validate internal address */
+      if((value&0x3f) != 0x09) {
+        /* fix internal address */
+        value &= ~0x3f;
+        value |= 0x09;
+      }
+      /* mask out reserved bits */
+      value &= 0x33ff;
+#ifdef CS8900_DEBUG_REGISTERS
+      log_message(cs8900_log,
+                  "Write TX Cmd  Register %04x: %04x",
+                  ppaddress,value);
+#endif
+    }
+    /* check tx_length register */
+    else if(ppaddress==0x146) {
+      /* HW always masks 0x0fff */
+      value &= 0x0fff;
+#ifdef CS8900_DEBUG_REGISTERS
+      log_message(cs8900_log,
+                  "Write TX Len  Register %04x: %04x",
+                  ppaddress,value);
+#endif
+    }
+    /* reserved range */
+    else if((ppaddress<0x144)||(ppaddress>0x147)) {
+#ifdef CS8900_DEBUG_WARN_REG
+      log_message(cs8900_log,
+                  "WARNING! Ignoring write to reserved Initiate Transmit Register %04x",
+                  ppaddress);
+#endif
+      return;
+    }
+  }
+
+  /* --- write to address filter register range --- */
+  else if(ppaddress<0x160) {
+    /* reserved range */
+    if(ppaddress>=0x15e) {
+#ifdef CS8900_DEBUG_WARN_REG
+      log_message(cs8900_log,
+                  "WARNING! Ingoring write to reserved Address Filter Register %04x",
+                  ppaddress);
+#endif
+      return;
+    }
+  }
+
+  /* --- ignore write outside --- */
+  else if(ppaddress<0x400) {
+#ifdef CS8900_DEBUG_WARN_REG
+    log_message(cs8900_log,
+                "WARNING! Ingoring write to reserved Register %04x",
+                ppaddress);
+#endif
+    return;
+  }
+  else if(ppaddress<0xa00){
+#ifdef CS8900_DEBUG_WARN_REG
+    log_message(cs8900_log,
+                "WARNING! Ignoring write to RX Buffer Range %04x",
+                ppaddress);
+#endif
+    return;
+  }
+  else {
+#ifdef CS8900_DEBUG_WARN_REG
+    log_message(cs8900_log,
+                "WARNIGN! Ignoring write to TX Buffer Range %04x",
+                ppaddress);
+#endif
+    return;
+  }
+
+  /* actually set value */
+  SET_PP_16(ppaddress, value);  
+}
+
+void cs8900a_device::cs8900_auto_incr_pp_ptr(void)
+{
+  /* perform auto increment of packet page pointer */
+  if((cs8900_packetpage_ptr & PP_PTR_AUTO_INCR_FLAG)==PP_PTR_AUTO_INCR_FLAG) {
+    /* pointer is always increment by one on real HW */
+    u16 ptr   = cs8900_packetpage_ptr & PP_PTR_ADDR_MASK;
+    u16 flags = cs8900_packetpage_ptr & PP_PTR_FLAG_MASK;
+    ptr++;
+    cs8900_packetpage_ptr = ptr | flags;
+  }
+}
+
+/* ----- read byte from I/O range ----- */
+u8 cs8900a_device::cs8900_read(u16 io_address)
```

Comment:
Suggest guarding against side-effects for debugger reads by testing machine().side_effects_disabled() in here.